In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('../Data/telecom_churn.csv')
df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

In [3]:
df.shape

(3333, 20)

**Notes**
- 20 features and 3333 observations. 
- Churn is the outcome
- Each observation is about a customer

In [4]:
df.isna().sum()

State                     0
Account length            0
Area code                 0
International plan        0
Voice mail plan           0
Number vmail messages     0
Total day minutes         0
Total day calls           0
Total day charge          0
Total eve minutes         0
Total eve calls           0
Total eve charge          0
Total night minutes       0
Total night calls         0
Total night charge        0
Total intl minutes        0
Total intl calls          0
Total intl charge         0
Customer service calls    0
Churn                     0
dtype: int64

**Notes**
- No nulls :) 

In [5]:
df.describe()

Account length    Area code  Number vmail messages  Total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       Total eve charge  Total night minutes  Total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       Total night charge  Total intl minutes  Total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       Total intl charge  Customer service calls  
count        3333.000000             3333.000000  
mean            2.764581                1.562856  
std             0.753773                1.315491  
min             0.000000                0.000000  
25%             2.300000                1.000000  
50%             2.780000                1.000000  
75%             3.270000                2.000000  
max             5.400000                9.000000

In [6]:
df.dtypes

State                      object
Account length              int64
Area code                   int64
International plan         object
Voice mail plan            object
Number vmail messages       int64
Total day minutes         float64
Total day calls             int64
Total day charge          float64
Total eve minutes         float64
Total eve calls             int64
Total eve charge          float64
Total night minutes       float64
Total night calls           int64
Total night charge        float64
Total intl minutes        float64
Total intl calls            int64
Total intl charge         float64
Customer service calls      int64
Churn                        bool
dtype: object

**Notes**
- 3 Objects 
- 1 Bool - The outcome 
- Int/Float - The rest

In [7]:
df.describe(include=['bool', 'object'])

State International plan Voice mail plan  Churn
count   3333               3333            3333   3333
unique    51                  2               2      2
top       WV                 No              No  False
freq     106               3010            2411   2850

**Notes**
- Lot of Customers don't have Interational plan. 3010/3333 do not have IP. 
- Churn is only seen in 483/3333 customer records. 

In [8]:
df['International plan'].value_counts(normalize = True)

No     0.90309
Yes    0.09691
Name: International plan, dtype: float64

In [9]:
df['Churn'].value_counts(normalize = True)

False    0.855086
True     0.144914
Name: Churn, dtype: float64

**Notes**
- 14.5% of customers are churning 

In [10]:
df.iloc[:3]

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False

In [11]:
df.iloc[[100,102],0:3]

State  Account length  Area code
100    IA              98        510
102    VT             135        415

In [12]:
# Since df['Churn'] is boolean, we can conver the dtype to int64
df['Churn'] = df['Churn'].astype('int64')
df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

In [13]:
df[(df['State'] == 'HI')  & (df['Churn'] == 1)]

State  Account length  Area code International plan Voice mail plan  \
1934    HI             105        415                 No              No   
2979    HI              99        408                 No              No   
3113    HI             108        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
1934                      0              281.3              124   
2979                      0              135.7              107   
3113                      0              291.6               99   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
1934             47.82              301.5               96             25.63   
2979             23.07              208.4              103             17.71   
3113             49.57              221.1               93             18.79   

      Total night minutes  Total night calls  Total night charge  \
1934                202.8                109                9.13   
2979                209.0                 95                9.40   
3113                229.2                110               10.31   

      Total intl minutes  Total intl calls  Total intl charge  \
1934                 8.7                 3               2.35   
2979                 8.8                 3               2.38   
3113                14.0                 9               3.78   

      Customer service calls  Churn  
1934                       0      1  
2979                       7      1  
3113                       2      1

In [14]:
df[(df['Churn'] == 1)  & (df['International plan'] == 'No')]['Customer service calls'].mean()

2.476878612716763

In [15]:
df[df['Churn'] == 1].mean() # 1 are True, those that left. 0 are those that are flase (stayed)

Account length            102.664596
Area code                 437.817805
Number vmail messages       5.115942
Total day minutes         206.914079
Total day calls           101.335404
Total day charge           35.175921
Total eve minutes         212.410145
Total eve calls           100.561077
Total eve charge           18.054969
Total night minutes       205.231677
Total night calls         100.399586
Total night charge          9.235528
Total intl minutes         10.700000
Total intl calls            4.163561
Total intl charge           2.889545
Customer service calls      2.229814
Churn                       1.000000
dtype: float64

In [16]:
df[df['Churn'] == 0].mean() 

Account length            100.793684
Area code                 437.074737
Number vmail messages       8.604561
Total day minutes         175.175754
Total day calls           100.283158
Total day charge           29.780421
Total eve minutes         199.043298
Total eve calls           100.038596
Total eve charge           16.918909
Total night minutes       200.133193
Total night calls         100.058246
Total night charge          9.006074
Total intl minutes         10.158877
Total intl calls            4.532982
Total intl charge           2.743404
Customer service calls      1.449825
Churn                       0.000000
dtype: float64

**Notes**
- When compared to 0's, 1's had higher average Customer Service calls 

In [17]:
df['Churn'].corr(df['Customer service calls'])

0.20874999878379413

In [23]:
charge = df[[i for i in df.columns if 'charge' in i]]
#[new candy  for candy in candy_jar if(conditions)]

In [24]:
charge.corr()

Total day charge  Total eve charge  Total night charge  \
Total day charge            1.000000          0.007036            0.004301   
Total eve charge            0.007036          1.000000           -0.012601   
Total night charge          0.004301         -0.012601            1.000000   
Total intl charge          -0.010094         -0.011074           -0.015186   

                    Total intl charge  
Total day charge            -0.010094  
Total eve charge            -0.011074  
Total night charge          -0.015186  
Total intl charge            1.000000